File: space-game/Core.Tests/Classes/DistanceRangeTests.vb
Imports NUnit.Framework
Imports Core.Classes

Namespace Classes
    <TestFixture>
    Public Class DistanceRangeTests

        <Test>
        Public Sub ctor_WithInvalidDefaultValues_ThrowsArgumentOutOfRangeException()
            Dim defaultLowerBound = Distance.FromAstronomicalUnits(10.4544)
            Dim defaultUpperBound = Distance.FromAstronomicalUnits(-50.4534)

            Dim expectedExceptionType As Type = GetType(ArgumentOutOfRangeException)

            Dim ex = Assert.Catch(Of Exception)(Function() New DistanceRange(defaultLowerBound, defaultUpperBound))

            Assert.AreEqual(ex.GetType(), expectedExceptionType)
        End Sub

        <Test>
        Public Sub LowerBound_WithDefaultValue_ReturnsDefaultValue()
            Dim defaultLowerBound = Distance.FromAstronomicalUnits(-10.3434)
            Dim defaultUpperBound = Distance.FromAstronomicalUnits(50.5465)
            Dim r = New DistanceRange(defaultLowerBound, defaultUpperBound)

            Dim result = r.LowerBound

            Assert.AreEqual(defaultLowerBound.Kilometers, result.Kilometers)
        End Sub

        <Test>
        Public Sub UpperBound_WithDefaultValue_ReturnsDefaultValue()
            Dim defaultLowerBound = Distance.FromAstronomicalUnits(-10.3434)
            Dim defaultUpperBound = Distance.FromAstronomicalUnits(50.7676)
            Dim r = New DistanceRange(defaultLowerBound, defaultUpperBound)

            Dim result = r.UpperBound

            Assert.AreEqual(defaultUpperBound.Kilometers, result.Kilometers)
        End Sub

        <Test>
        Public Sub LowerBound_WithValues_ReturnsCorrectValue()
            Dim defaultLowerBound = Distance.FromAstronomicalUnits(-10.7767)
            Dim defaultUpperBound = Distance.FromAstronomicalUnits(50.3434)
            Dim r = New DistanceRange(defaultLowerBound, defaultUpperBound)

            Dim expectedValue = Distance.FromAstronomicalUnits(-20.2323)
            r.AddValue(Distance.FromAstronomicalUnits(50.767))
            r.AddValue(Distance.FromAstronomicalUnits(-12.5656))
            r.AddValue(Distance.FromAstronomicalUnits(17.6565))
            r.AddValue(expectedValue)
            r.AddValue(Distance.FromAstronomicalUnits(0))
            r.AddValue(Distance.FromAstronomicalUnits(13))

            Dim result = r.LowerBound

            Assert.AreEqual(expectedValue.Kilometers, result.Kilometers)
        End Sub

        <Test>
        Public Sub UpperBound_WithValues_ReturnsCorrectValue()
            Dim defaultLowerBound = Distance.FromAstronomicalUnits(-10)
            Dim defaultUpperBound = Distance.FromAstronomicalUnits(50.656)
            Dim r = New DistanceRange(defaultLowerBound, defaultUpperBound)

            Dim expectedValue = Distance.FromAstronomicalUnits(200.878)
            r.AddValue(Distance.FromAstronomicalUnits(50))
            r.AddValue(Distance.FromAstronomicalUnits(-12))
            r.AddValue(Distance.FromAstronomicalUnits(17.878))
            r.AddValue(expectedValue)
            r.AddValue(Distance.FromAstronomicalUnits(0))
            r.AddValue(Distance.FromAstronomicalUnits(13))

            Dim result = r.UpperBound

            Assert.AreEqual(expectedValue.Kilometers, result.Kilometers)
        End Sub

    End Class
End Namespace

File: space-game/Core.Tests/Classes/DistanceRangeTests.cs
using Core.Classes;
using NUnit.Framework;
using System;

namespace Core.Tests.Classes
{
    [TestFixture]
    public class DistanceRangeTests
    {
        [Test]
        public void ctor_WithInvalidDefaultValues_ThrowsArgumentOutOfRangeException()
        {
            Distance defaultLowerBound = Distance.FromAstronomicalUnits(10.4544);
            Distance defaultUpperBound = Distance.FromAstronomicalUnits(-50.4534);

            Type expectedExceptionType = typeof(ArgumentOutOfRangeException);

            Exception ex = Assert.Catch<Exception>(() => new DistanceRange(defaultLowerBound, defaultUpperBound));

            Assert.AreEqual(ex.GetType(), expectedExceptionType);
        }

        [Test]
        public void LowerBound_WithDefaultValue_ReturnsDefaultValue()
        {
            Distance defaultLowerBound = Distance.FromAstronomicalUnits(-10.3434);
            Distance defaultUpperBound = Distance.FromAstronomicalUnits(50.5465);
            DistanceRange r = new DistanceRange(defaultLowerBound, defaultUpperBound);

            Distance result = r.LowerBound;

            Assert.AreEqual(defaultLowerBound.Kilometers, result.Kilometers);
        }

        [Test]
        public void UpperBound_WithDefaultValue_ReturnsDefaultValue()
        {
            Distance defaultLowerBound = Distance.FromAstronomicalUnits(-10.3434);
            Distance defaultUpperBound = Distance.FromAstronomicalUnits(50.7676);
            DistanceRange r = new DistanceRange(defaultLowerBound, defaultUpperBound);

            Distance result = r.UpperBound;

            Assert.AreEqual(defaultUpperBound.Kilometers, result.Kilometers);
        }

        [Test]
        public void LowerBound_WithValues_ReturnsCorrectValue()
        {
            Distance defaultLowerBound = Distance.FromAstronomicalUnits(-10.7767);
            Distance defaultUpperBound = Distance.FromAstronomicalUnits(50.3434);
            DistanceRange r = new DistanceRange(defaultLowerBound, defaultUpperBound);

            Distance expectedValue = Distance.FromAstronomicalUnits(-20.2323);
            r.AddValue(Distance.FromAstronomicalUnits(50.767));
            r.AddValue(Distance.FromAstronomicalUnits(-12.5656));
            r.AddValue(Distance.FromAstronomicalUnits(17.6565));
            r.AddValue(expectedValue);
            r.AddValue(Distance.FromAstronomicalUnits(0));
            r.AddValue(Distance.FromAstronomicalUnits(13));

            Distance result = r.LowerBound;

            Assert.AreEqual(expectedValue.Kilometers, result.Kilometers);
        }

        [Test]
        public void UpperBound_WithValues_ReturnsCorrectValue()
        {
            Distance defaultLowerBound = Distance.FromAstronomicalUnits(-10);
            Distance defaultUpperBound = Distance.FromAstronomicalUnits(50.656);
            DistanceRange r = new DistanceRange(defaultLowerBound, defaultUpperBound);

            Distance expectedValue = Distance.FromAstronomicalUnits(200.878);
            r.AddValue(Distance.FromAstronomicalUnits(50));
            r.AddValue(Distance.FromAstronomicalUnits(-12));
            r.AddValue(Distance.FromAstronomicalUnits(17.878));
            r.AddValue(expectedValue);
            r.AddValue(Distance.FromAstronomicalUnits(0));
            r.AddValue(Distance.FromAstronomicalUnits(13));

            Distance result = r.UpperBound;

            Assert.AreEqual(expectedValue.Kilometers, result.Kilometers);
        }
    }
}
